### PR TITLE
[SPARK-15701][SQL] Modify ColumnVector to reduce memory usage for constant vectors

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

For `ColumnVector` used as the partition columns and null columns, because they only include one distinct value, we don't need to allocate full capacity for the number of rows, but just 1 row.

This can reduce its memory usage and from benchmark results it also speeds up data access.
## How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by kiszk** [comment]
Can we have a benchmark program to show performance improvement?

**Comment by viirya** [comment]
ColumnVector R/W:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    On Heap, Not Constant                           91 /   93          0.4        2224.7       1.0X   
    On Heap, Constant                               46 /   46          0.9        1120.7       2.0X   
    Off Heap, Not Constant                        1127 / 1295          0.0       27503.4       0.1X   
    Off Heap, Constant                             969 / 1008          0.0       23662.3       0.1X

**Comment by viirya** [comment]
@kiszk I've run some benchmark codes. The benchmark results:

    ColumnVector R/W:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    On Heap, Not Constant                           57 /   59          0.7        1395.4       1.0X
    On Heap, Constant                               44 /   44          0.9        1062.0       1.3X
    Off Heap, Not Constant                        1059 / 1092          0.0       25861.4       0.1X
    Off Heap, Constant                             998 / 1143          0.0       24373.4       0.1X

**Comment by viirya** [comment]
Don't know who is best person to review this... cc @rxin

**Comment by rxin** [comment]
Sorry I don't understand what this is trying to accomplish from your pr description. Are you trying to save space for the partition column?

This adds the number of concrete implementations for column batch from 2 to 3, which would eliminate JIT inlining, and can be pretty bad for performance.

**Comment by viirya** [comment]
For the partition columns and null columns, we don't need to allocate full capacity for the number of rows. Because there is only one value in these columns. Yes, we only need allocate the space for 1 row.

I have another implementation that doesn't change the number of concrete implementation. If this is the concern you have, I will change it to the implementation.

**Comment by kiszk** [comment]
@viirya , thank you for preparing benchmark results. Let me clarify this implementation by two questions.

1.  To reduce memory foot print, does this PR allocate an array for each column with the number of row instead of ``` DEFAULT_BATCH_SIZE```?
2.  Why can this PR improve performance?  When I see ```getInt()``` both for ```OnHeapColumnVector.java``` and ```OnHeapConstantColumnVector.java```, they are almost similar (```intData[rowId]``` and ```intData[0]```). Does the performance improvement come from possible elimination of index bound check due to constant index access? Correct?
If so, we could do the similar thing by using ```Platform.getInt()``` in ```OnHeapColumnVector.java```.

**Comment by viirya** [comment]
@kiszk 

1. The allocation mechanism is not changed in this patch. Just for the column vector representing partition columns and null columns, this patch only allocates 1 row.

2. I think the speed-up might come from simplified memory access. Because we don't need to scan the memory space for all values. For the constant column vectors, only one value is accessed. I don't know if elimination of index bound helps or not. I will try to do that in `OnHeapColumnVector` and benchmark them again to see the difference.

**Comment by viirya** [comment]
@kiszk I've tried elimination of index bound in  `OnHeapColumnVector` by using `Platform.getInt()`. However, I do not see performance improvement.

**Comment by rxin** [comment]
Can you talk about your other implementation here? The current implementation seems complicated for saving some memory. Perhaps one way to do it is to just do it in code generation.

**Comment by viirya** [comment]
@rxin Another implementation is to check if the column vector is constant and do corresponding logic in element access. Do it in code generation sounds interesting. As the column vector is not code generation now, do you mean to move it to code generation at all?

**Comment by rxin** [comment]
Can you tell me what you are trying to address here? Are you seeing a problem with memory usage? I just don't know why this is important ... the number of partition columns is typically small.

**Comment by viirya** [comment]
@rxin You are right. This is an improvement to address the memory usage. I just think that if we can reduce the memory usage with a change not too big, it might be worth doing this. BTW, from the benchmark, looks like it also benefits access performance a bit.

**Comment by rxin** [comment]
But how much performance can we gain in reality with this? The number of partition columns is usually not that large in practice.

**Comment by viirya** [comment]
Yea. That is right. The problem is whether it is worth doing this change for the memory usage reduction and performance gain. If the change is relatively small (another implementation, not current one), I think it might be worth. Current change is complicated as you said. Code generation seems too much for this. So you think this patch is not worth doing? Or let me update this to another implementation and then you take a look if it is worth?

**Comment by kiszk** [comment]
Can we implement this feature simply by 

- using ```ColumnarBatch.allocate(StructType schema, MemoryMode memMode, int maxRows)``` with ```maxRows=1```
- not introducing ```OffHeapConstantColumnVector``` and ```OnHeapConstantColumnVector```

IMHO, if this PR focuses on memory usage reduction, it could be possible.

**Comment by viirya** [comment]
@kiszk Yea. I am going to remove `OffHeapConstantColumnVector` and `OnHeapConstantColumnVector`.

**Comment by viirya** [comment]
@kiszk BTW, we can't simply do this by `ColumnarBatch.allocate` with `maxRows=1`. Because we still need to take care of element access. In other words, from outside, the vector looks like it has the same number of elements as other columns, not just 1 row.

**Comment by kiszk** [comment]
Oh, you are right. IMHO, it is too complex to introduce new implementation classes only for a column vector with the same value in all of the rows.
To introduce compression schemes, as implemented in ```CachedBatch``` may be more generic solution if we introduce new implementation classes.

**Comment by viirya** [comment]
@kiszk Yea. As @rxin said, only for reducing memory usage the current change is too much. I will change this to simple implementation.

**Comment by viirya** [comment]
Benchmark again on new change:

Environment:

    Java HotSpot(TM) 64-Bit Server VM 1.8.0_71-b15 on Linux 3.19.0-25-generic
    Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz

OnHeap, Not Constant:

    ColumnVector R/W:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    On Heap, Not Constant                           42 /   49          1.0        1020.3       1.0X

    ColumnVector R/W:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    On Heap, Not Constant                           41 /   46          1.0         989.0       1.0X

OnHeap, Constant:

    ColumnVector R/W:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    On Heap, Constant                               28 /   33          1.5         674.2       1.0X   

    ColumnVector R/W:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    On Heap, Constant                               27 /   33          1.5         658.4       1.0X   

OffHeap, Not Constant:

    ColumnVector R/W:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    Off Heap, Not Constant                          63 /   73          0.6        1547.3       1.0X   

    ColumnVector R/W:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    Off Heap, Not Constant                          68 /   74          0.6        1663.5       1.0X   

OffHeap, Constant:

    ColumnVector R/W:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    Off Heap, Constant                              27 /   33          1.5         662.5       1.0X

    ColumnVector R/W:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    Off Heap, Constant                              27 /   33          1.5         657.1       1.0X

**Comment by viirya** [comment]
The latest benchmark is run individually for each type of column vector. As stated in `ColumnarBatchBenchmark`, it is hard to reason about the JIT. If we put these 4 cases together to run benchmark, the numbers seems not accurate and looks weird.

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OnHeapColumnVector.java [hunk 1]
```diff
@@ -70,26 +71,106 @@ public long nullsNativeAddress() {
   public void close() {
   }
 
+  // Spilt this function out since it is the slow path.
```
Comment (viirya) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OnHeapColumnVector.java]: Not changed. Just add back the codes moved to new classes in previous commit.

**Comment by viirya** [comment]
@rxin I've updated this to more simple approach that doesn't introduce new classes. The main change is to check if the current vector is constant or not and do suitable data access. Please take a look. Thanks.

**Comment by rxin** [comment]
@viirya this is still a pretty major change for unclear benefits. There might be other more important things that need more eyes on...

**Comment by viirya** [comment]
@rxin hmm, I just think if we can improve it by just adding conditional check, it might be worth doing.

For the performance hurt, this is benchmark for on-heap and off-heap column vectors before this patch:

On Heap:

    ColumnVector R/W:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    On Heap                                         39 /   47          1.1         946.8       1.0X

    ColumnVector R/W:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    On Heap                                         41 /   46          1.0         995.5       1.0X

Off Heap:

    ColumnVector R/W:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    Off Heap                                        65 /   75          0.6        1598.2       1.0X

    ColumnVector R/W:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    Off Heap                                        63 /   74          0.7        1532.5       1.0X 

Looks like the performance is not hurt obviously/significantly.

But if you still have concerns about this, we can close this.

**Comment by rxin** [comment]
I am not sure if you are really testing it correctly -- your benchmark is mostly likely just testing how well the CPU does branch prediction when the flag is always true or false.

**Comment by viirya** [comment]
hmm, but as the flag is set, I think it will not be changed?

**Comment by viirya** [comment]
Besides, I just wrote this test according to other tests in `ColumnarBatchBenchmark` that benchmark on-heap, off-heap column vector access. I was thinking it might be enough. If not, any else need to test further?

**Comment by rxin** [comment]
What I meant is that if in one process you have some invocation of the function that would hit the true branch, and some other invocation of the function that would hit the false branch, the performance is going to be worse. Google "branch prediction" for more information.

Basically you can't measure the overhead of an extra branch in practice by running a benchmark in which the flag is either always false or always true.

**Comment by viirya** [comment]
I see. My question is, as for example we create 2 column vectors, one is constant and one is not. Because we will not re-use the column vectors, so their constant flag is fixed and not changed. As they are two different instances, will the problem you said happen? When `getInt` of first vector (constant) is called and later  `getInt` of the second (not constant) is called, the performance will be down?

**Comment by viirya** [comment]
So will it be more practice to benchmark the case in which there are some constant and some not constant column vectors are used together? And compare it with the original case in which all columns are not with this extra branch (i.e., without this path)?

**Comment by rxin** [comment]
Going back to my original question: what's the point of this complicated pull request? How much memory would you save in practice? The column batches are not for persistent memory storage yet, and they are supposed to be only for a small number of rows.

**Comment by viirya** [comment]
Ok. I got it. So I think the point is the memory usage reduction is not worth doing this change. Let me close it now.

**Pull request closed without merging**